languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/SessionFactoryTest.java

```diff
@@ -42,32 +49,677 @@
+        .build());
+    when(policy.getSessionCacheExpireMillis()).thenReturn(sessionCacheExpireMillis);
+
+    try (SessionFactory factory = SessionFactory.newBuilder(testProductId, testServiceId)
+        .withMetastore(metastoreSpy)
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+        byte[] payload = new byte[] {0, 1, 2, 3, 4, 5, 6, 7};
+        byte[] drr = session.encrypt(payload);
+
+        // Sleep to trigger the expiration
+        // TODO Consider refactoring the Caffeine usage to allow injecting a ticker if sleeping becomes unreliable
+        try {
+          Thread.sleep(sessionCacheExpireMillis * 3);
+        } catch (Exception e) {
+          fail(e.getMessage());
+        }
+
+        // Reset so we can examine 2nd session's interactions
+        reset(metastoreSpy);
+
+        // Use same partition to get the same cached (but expired) session while it's still in use
+        try (Session<byte[], byte[]> session2 = factory.getSessionBytes(testPartitionId)) {
+          byte[] decryptedPayload = session2.decrypt(drr);
+
+          assertArrayEquals(payload, decryptedPayload);
+
+          // verify no metastore interactions in the decrypt flow (since IKs cached via session caching)
+          verify(metastoreSpy, never()).load(any(), any());
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  void testSessionCacheGetSessionAfterUseAndNotExpiredShouldNotEvict() {
+    CryptoPolicy policy = BasicExpiringCryptoPolicy.newBuilder()
+        .withKeyExpirationDays(1)
+        .withRevokeCheckMinutes(30)
+        .withCanCacheSessions(true)
+        .build();
+
+    try (SessionFactory factory = SessionFactory.newBuilder(testProductId, testServiceId)
+        .withMetastore(metastoreSpy)
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      byte[] payload = new byte[] {0, 1, 2, 3, 4, 5, 6, 7};
+      byte[] drr = null;
+      try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+        drr = session.encrypt(payload);
+        byte[] decryptedPayload = session.decrypt(drr);
+
+        assertArrayEquals(payload, decryptedPayload);
+      } catch (Exception e) {
+        fail(e.getMessage());
+      }
+
+      // Note we do not sleep
+
+      // Reset so we can examine 2nd session's interactions
+      reset(metastoreSpy);
+
+      try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+        byte[] decryptedPayload = session.decrypt(drr);
+
+        assertArrayEquals(payload, decryptedPayload);
+        // verify no metastore interactions in the decrypt flow (since IKs cached via session caching)
+        verify(metastoreSpy, never()).load(any(), any());
+      } catch (Exception e) {
+        fail(e.getMessage());
+      }
```

Comment:
Can verify before closing the outer session that nothing is evicted from cache since the session is still in use.

Reply:
similar comment as above. can add the extra check